File: app/src/androidTest/java/com/quickhandslogistics/resources/FetchingIdlingResource.java
package com.quickhandslogistics.resources;

import androidx.test.espresso.IdlingResource;

import okhttp3.Dispatcher;

public class FetchingIdlingResource implements IdlingResource {

    private IdlingResource.ResourceCallback resourceCallback;

    public FetchingIdlingResource(Dispatcher okHttpClientDispatcher) {
        okHttpClientDispatcher.setIdleCallback(() -> {
            if (resourceCallback != null) {
                resourceCallback.onTransitionToIdle();
            }
        });
    }

    @Override
    public String getName() {
        return FetchingIdlingResource.class.getSimpleName();
    }

    @Override
    public boolean isIdleNow() {
        return false;
    }

    @Override
    public void registerIdleTransitionCallback(ResourceCallback callback) {
        resourceCallback = callback;
    }
}

File: app/src/androidTest/java/com/quickhandslogistics/resources/FetchingIdlingResource.kt
package com.quickhandslogistics.resources

import androidx.test.espresso.IdlingResource
import okhttp3.Dispatcher

class FetchingIdlingResource(private val okHttpClientDispatcher: Dispatcher?) : IdlingResource {
    private var resourceCallback: IdlingResource.ResourceCallback? = null

    init {
        okHttpClientDispatcher?.setIdleCallback {
            resourceCallback?.let { callback ->
                callback.onTransitionToIdle()
            }
        }
    }

    override fun getName(): String {
        return FetchingIdlingResource::class.java.simpleName
    }

    override fun isIdleNow(): Boolean {
        return okHttpClientDispatcher?.runningCallsCount() == 0
    }

    override fun registerIdleTransitionCallback(callback: IdlingResource.ResourceCallback?) {
        resourceCallback = callback
    }
}